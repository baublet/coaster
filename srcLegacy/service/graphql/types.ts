import { ServiceType, ServiceDefaultProperties } from "../types";
import { Controller } from "../controller/types";
import { ModelFactory } from "model";
import { PersistedModelFactory } from "persist";
import { PersistQueryBuilder } from "persist/types";
import { ModelConnectionInputArguments } from "./modelConnectionResolvers";

export type GraphQLPrimitiveTypes = {
  type: GraphQLPrimitiveType;
  nullable?: boolean;
  resolver?: GraphQLTypedResolverDeclaration;
};

export type GraphQLModelType = {
  type: GraphQLType.MODEL;
  modelFactory: ModelFactory;
  nullable?: boolean;
  description?: string;
  resolver?: GraphQLTypedResolverDeclaration;
};

export type GraphQLModelConnectionType = {
  type: GraphQLType.MODEL_CONNECTION;
  /**
   * Pass in a name here to override the <Model>ModelConnection default name
   */
  name?: string;
  /**
   * Connections allow constraints for pagination and searching. By default, we
   * name that input <Model>ModelConnectionInput. Override the default with this
   */
  inputName?: string;
  modelFactory: PersistedModelFactory;
  nullable?: boolean;
  description?: string;
};

export type GraphQLConnectionType = {
  type: GraphQLType.CONNECTION;
  name?: string;
  of: GraphQLReturnStructureNode;
  nullable?: boolean;
  description?: string;
  resolver?: GraphQLTypedResolverDeclaration;
};

export type GraphQLArrayOfType = {
  type: GraphQLType.ARRAY_OF;
  value: GraphQLReturnStructureNode;
  nullable?: boolean;
  resolver?: GraphQLTypedResolverDeclaration;
};

export type GraphQLUnionType = {
  type: GraphQLType.UNION;
  values: GraphQLObjectType[];
  nullable?: boolean;
  resolver?: GraphQLTypedResolverDeclaration;
};

export type GraphQLObjectType = {
  type: GraphQLType.OBJECT;
  nodes: Record<string, GraphQLReturnStructureNode>;
  nullable?: boolean;
  name?: string;
  description?: string;
  resolver?: GraphQLTypedResolverDeclaration;
};

type GraphQLEnumFieldsMap = Record<
  string,
  | string
  | {
      deprecationReason?: string;
      description: string;
    }
>;

export type GraphQLEnumType = {
  type: GraphQLType.ENUM;
  nullable?: boolean;
  name?: string;
  description?: string;
  values: GraphQLEnumFieldsMap;
  resolver?: GraphQLTypedResolverDeclaration;
};

export enum GraphQLType {
  ARRAY_OF = "ARRAY_OF",
  BOOLEAN = "BOOLEAN",
  CONNECTION = "CONNECTION",
  ENUM = "ENUM",
  FLOAT = "FLOAT",
  ID = "ID",
  INT = "INT",
  MODEL = "MODEL",
  MODEL_CONNECTION = "MODEL_CONNECTION",
  OBJECT = "OBJECT",
  SCALAR = "SCALAR",
  STRING = "STRING",
  UNION = "UNION"
}

export type GraphQLPrimitiveType =
  | GraphQLType.SCALAR
  | GraphQLType.BOOLEAN
  | GraphQLType.STRING
  | GraphQLType.FLOAT
  | GraphQLType.ID
  | GraphQLType.INT;

export type GraphQLReturnStructureNode =
  | GraphQLPrimitiveTypes
  | GraphQLModelType
  | GraphQLModelConnectionType
  | GraphQLConnectionType
  | GraphQLArrayOfType
  | GraphQLObjectType
  | GraphQLEnumType
  | GraphQLUnionType;

// We omit models and model collections here. They don't make sense as arguments
export type GraphQLResolverArgument =
  | GraphQLPrimitiveTypes
  | GraphQLArrayOfType
  | GraphQLObjectType;

export type GraphQLResolverArguments = Record<string, GraphQLResolverArgument>;

export type ArgumentTypeFromArguments<
  T extends GraphQLResolverArguments
> = OptionalNodes<GraphQLObjectDeclarationTypes<T>, RequiredKeysInNodes<T>> &
  RequiredNodes<GraphQLObjectDeclarationTypes<T>, RequiredKeysInNodes<T>>;

export type GraphQLReturnStructureRootNode = GraphQLReturnStructureNode & {
  typename?: string;
};

export interface GraphQLQueryControllerConfiguration<
  ResolverArguments extends GraphQLResolverArguments = any,
  ReturnStructure extends GraphQLReturnStructureNode = any
> {
  description?: string;
  resolver: Controller<ResolverArguments, ReturnStructure>;
  resolverArguments?: GraphQLResolverArguments;
  resolutionType: GraphQLReturnStructureRootNode;
}

export type GraphQLQueryOrMutationNode = {
  [key: string]: GraphQLQueryControllerConfiguration;
};

export interface GraphQLServiceOptions {
  queries: GraphQLQueryOrMutationNode;
  mutations?: GraphQLQueryOrMutationNode;
}

export interface GraphQLServiceArguments extends ServiceDefaultProperties {
  type: ServiceType.GRAPHQL;
  options: GraphQLServiceOptions;
}

export type GraphQLPrimitiveEnumToType<
  T extends GraphQLPrimitiveType
> = T extends GraphQLType.STRING
  ? string
  : T extends GraphQLType.SCALAR
  ? any
  : T extends GraphQLType.FLOAT
  ? number
  : T extends GraphQLType.INT
  ? number
  : T extends GraphQLType.ID
  ? string
  : T extends GraphQLType.BOOLEAN
  ? boolean
  : never;

export type GraphQLObjectDeclarationNodes = Record<
  string,
  GraphQLReturnStructureNode
>;

export type GraphQLObjectDeclarationTypes<
  T extends GraphQLObjectDeclarationNodes
> = {
  [K in keyof T]: ReturnNodeToType<T[K]>;
};

export type GraphQLObjectDeclarationToType<
  T extends GraphQLObjectDeclarationNodes
> = RequiredNodes<GraphQLObjectDeclarationTypes<T>, RequiredKeysInNodes<T>> &
  OptionalNodes<GraphQLObjectDeclarationTypes<T>, RequiredKeysInNodes<T>>;

export type RequiredKeysInNodes<T extends GraphQLObjectDeclarationNodes> = {
  [K in keyof T]: T[K]["nullable"] extends false ? K : never;
}[keyof T];

export type OptionalNodes<
  T extends GraphQLObjectDeclarationNodes,
  K extends keyof T
> = Partial<Omit<T, K>>;

export type RequiredNodes<
  T extends GraphQLObjectDeclarationNodes,
  K extends keyof T
> = Pick<T, K>;

export type ReturnNodeToType<
  T extends GraphQLReturnStructureNode
> = T extends GraphQLPrimitiveTypes
  ? GraphQLPrimitiveEnumToType<T["type"]>
  : T extends GraphQLObjectType
  ? GraphQLObjectDeclarationToType<T["nodes"]>
  : T extends GraphQLArrayOfType
  ? ReturnNodeToType<T["value"]>[]
  : T extends GraphQLEnumType
  ? keyof T["values"]
  : never;

export type ReturnNodeToTypeOptionality<
  T extends any
> = T["nullable"] extends false ? undefined : T;

export type GraphQLTypedResolverDeclaration<
  ResolverArguments extends GraphQLResolverArguments = any,
  ReturnTypeArgument extends GraphQLReturnStructureNode = any
> = GraphQLQueryControllerConfiguration<
  ArgumentTypeFromArguments<ResolverArguments>,
  ReturnNodeToTypeOptionality<ReturnNodeToType<ReturnTypeArgument>>
>;

export type GraphQLModelConnectionResolver = (
  parent: any,
  input: ModelConnectionInputArguments,
  context: any,
  info: any
) => Promise<{
  queryWithBaseConstraints: PersistQueryBuilder;
  input: ModelConnectionInputArguments;
}>;
