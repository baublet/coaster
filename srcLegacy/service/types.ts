import {
  GraphQLServiceArguments,
  GraphQLServiceOptions
} from "./graphql/types";
import { RESTServiceArguments, RESTServiceOptions } from "./rest/types";
import DataLoader from "dataloader";
import { PersistedModel } from "persist/types";

export interface Service {
  name: string;
  host: string;
  port: number;
  protocol: string;
  initialize: () => Promise<void>;
  start: () => Promise<void>;
}

export enum ServiceType {
  GRAPHQL,
  REST,
  WEB
}

export type ServiceArguments = GraphQLServiceArguments | RESTServiceArguments;

export interface ServiceDefaultProperties {
  name: string;
  host?: string;
  port: number;
  type: ServiceType;
  options: GraphQLServiceOptions | RESTServiceOptions;
}

export interface ServiceContext {
  modelDataLoaders: Record<string, DataLoader<string, PersistedModel>>;
}
