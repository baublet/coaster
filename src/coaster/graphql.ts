export { createGraphqlEndpointHandler } from "./src/graphql/createGraphqlEndpointHandler";
export { createGraphqlTrack } from "./src/graphql/createGraphqlTrack";
