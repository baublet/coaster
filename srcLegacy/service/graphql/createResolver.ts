import {
  GraphQLResolverArguments,
  GraphQLTypedResolverDeclaration,
  GraphQLReturnStructureNode,
  ReturnNodeToType,
  ArgumentTypeFromArguments,
  ReturnNodeToTypeOptionality
} from "./types";
import { Controller } from "service/controller/types";

/**
 * Creates a GraphQL resolver from a specified resolver configuration. Enforces
 * strongly-typed resolvers.
 */
export function createResolver<
  ResolverArguments extends GraphQLResolverArguments,
  ReturnNode extends GraphQLReturnStructureNode
>({
  description,
  resolutionType,
  resolverArguments,
  resolver
}: {
  description?: string;
  resolverArguments?: ResolverArguments;
  resolutionType: ReturnNode;
  resolver: Controller<
    ArgumentTypeFromArguments<ResolverArguments>,
    ReturnNodeToTypeOptionality<ReturnNodeToType<ReturnNode>>
  >;
}): GraphQLTypedResolverDeclaration<ResolverArguments, ReturnNode> {
  return {
    description,
    resolver,
    resolutionType,
    resolverArguments
  };
}
