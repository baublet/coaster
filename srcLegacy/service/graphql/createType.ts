import { GraphQLReturnStructureNode } from "./types";

export function createType<T extends GraphQLReturnStructureNode>(
  typeDef: T
): T {
  return typeDef;
}
