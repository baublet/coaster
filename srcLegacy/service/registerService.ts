import { ServiceArguments, Service, ServiceType } from "./types";

export function registerService<Args extends ServiceArguments>(
  args: Args
): Service {
  return (args as any) as Service;
}

// Type tests

export const service: Service = registerService({
  name: "Test service",
  type: ServiceType.GRAPHQL,
  port: 80,
  host: "0.0.0.0",
  options: {
    resolvers: {}
  }
});
