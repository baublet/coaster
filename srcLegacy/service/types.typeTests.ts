import { ServiceArguments, ServiceType } from "./types";

// export const gqlService: ServiceArguments = {
//   name: "GraphQL Service",
//   port: 80,
//   type: ServiceType.GRAPHQL,
//   options: {
//     resolvers: {}
//   }
// };

export const restService: ServiceArguments = {
  name: "REST Service",
  port: 80,
  type: ServiceType.REST,
  options: {
    routes: {}
  }
};
