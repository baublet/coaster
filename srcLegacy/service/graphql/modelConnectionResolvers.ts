import {
  GraphQLInputObjectType,
  GraphQLInt,
  GraphQLList,
  GraphQLEnumType,
  GraphQLString,
  GraphQLNonNull
} from "graphql";
import { PersistedModelFactory } from "persist";
import { ServiceContext } from "service/types";
import {
  PersistModelRelationship,
  PersistedModel,
  PersistQueryBuilder
} from "persist/types";
import { createDataLoaderForModelFactory } from "./createDataLoaderForModelFactory";

export interface ModelListResolverParent {
  queryWithBaseConstraints: PersistQueryBuilder;
  input: ModelConnectionInputArguments;
}

export interface ModelConnectionInputArguments {
  filter?: string;
  limit?: number;
  offset?: number;
  sort?: {
    field: string;
    direction: "ASC" | "DESC";
  }[];
}

const SortDirection = new GraphQLEnumType({
  name: "SortDirection",
  values: {
    ASC: {
      description: "Ascending"
    },
    DESC: {
      description: "Descending"
    }
  }
});

const SortOptions = new GraphQLInputObjectType({
  name: "SortOptions",
  fields: {
    field: {
      type: new GraphQLNonNull(GraphQLString),
      description: "Field to sort on"
    },
    direction: {
      type: SortDirection,
      description: "Direction to sort by",
      defaultValue: "ASC"
    }
  }
});

/**
 * Returns a resolver for default model connections when they're associated
 * with models they have relationships to.
 */
export function defaultRelationshipFactory() {
  // parentFactory: PersistedModelFactory,
  // childFactory: PersistedModelFactory
  // const parentPrimaryKey = parentFactory.$options.primaryKey;
  // const childPrimaryKey = childFactory.$options.primaryKey;

  return (): // parent: PersistedModelFactory,
  // input: ModelConnectionInputArguments,
  // context: ServiceContext
  Promise<PersistQueryBuilder> => {
    // TODO: filter along input.filter
    return 1 as any;
  };
}

export function connectionInput(factory: PersistedModelFactory) {
  return new GraphQLInputObjectType({
    name: `${factory.name}ModelConnectionInput`,
    description: `Filtering, sorting, and pagination for ${factory.name}`,
    fields: {
      filters: {
        type: GraphQLString,
        description:
          "Filter connection results by these options. Uses the parser format from the following module: https://www.npmjs.com/package/search-query-parser"
      },
      limit: {
        type: GraphQLInt,
        description: "The maximum number of results to return"
      },
      offset: {
        type: GraphQLInt,
        description: "Skip this number of results before returning results"
      },
      sort: {
        type: new GraphQLList(new GraphQLNonNull(SortOptions))
      }
    }
  });
}

export interface ModelConnectionRootNodeReturn<
  Parent extends PersistedModelFactory,
  Child extends PersistedModelFactory
> {
  parent: ReturnType<Parent>;
  parentFactory: Parent;
  connectionFactory: Child;
  input: ModelConnectionInputArguments;
  relationship: PersistModelRelationship;
}

/**
 * Creates a resolver for `totalCount` on a connection type. The parent should
 * always be an object with one property: queryWithBaseConstraints. That prop
 * is an instance `PersistQueryBuilder` with the initial constraints (like the
 * whereIn, any filters applied, but _never_ with limit and offsets applied).
 */
export function totalCountResolver() {
  return async ({
    queryWithBaseConstraints
  }: ModelListResolverParent): Promise<number> => {
    return await queryWithBaseConstraints.count();
  };
}

export interface ModelConnectionNodeNotLoadedValue<
  Child extends PersistedModelFactory = any
> {
  id: string;
  connectionFactory: Child;
}

/**
 * Creates a resolver for `nodes` on a connection type. The parent should
 * always be an object with a property queryWithBaseConstraints. That prop
 * is an instance `PersistQueryBuilder` with the initial constraints (like the
 * whereIn, any filters applied, but _never_ with limit and offsets applied).
 *
 * The resolver also takes the initial input from the top-level connection
 * resolver so that we can correctly limit, offset, and order the result-set.
 *
 * It resolves into an array of IDs that field-level resolvers handle resolving
 * via a dataLoader.
 */
export function modelListResolverFactory<T extends PersistedModelFactory>(
  connectionModelFactory: T
) {
  const childPrimaryKey = connectionModelFactory.$options.primaryKey;
  return async ({
    queryWithBaseConstraints,
    input
  }: ModelListResolverParent): Promise<string[]> => {
    if (input.limit) queryWithBaseConstraints.limit(input.limit);
    if (input.offset) queryWithBaseConstraints.offset(input.offset);
    if (input.sort) {
      for (const { field, direction = "ASC" } of input.sort) {
        queryWithBaseConstraints.orderBy(field, direction);
      }
    }

    const results = await queryWithBaseConstraints;

    return results.map(result => ({
      id: result[childPrimaryKey],
      connectionFactory: connectionModelFactory
    }));
  };
}

/**
 * A default resolver for loading models in any connection. If you're a parent
 * returning models, return or resolve from the parent a string that represents
 * the model's primary key to load via DataLoader.
 */
export function nodeResolverFactory<T extends PersistedModelFactory>(
  modelFactory: T
) {
  return async (
    id: string,
    _args: unknown,
    context: ServiceContext
  ): Promise<PersistedModel> => {
    // Some basic runtime error checking
    if (typeof id !== "string") {
      throw new Error(
        `We tried to resolve a model without an ID. Coaster ships with default resolvers that utilize dataLoaders for models you utilize in your GraphQL schema. These resolvers take as a parent a simple ID. Make sure your parent resolver returns a string of the model you want to load!`
      );
    }

    // Create the dataloader if it doesn't exist
    if (!context.modelDataLoaders) {
      context.modelDataLoaders = {};
    }
    if (!context.modelDataLoaders[modelFactory.name]) {
      context.modelDataLoaders[
        modelFactory.name
      ] = createDataLoaderForModelFactory(modelFactory);
    }

    const dataLoader = context.modelDataLoaders[modelFactory.name];
    return dataLoader.load(id);
  };
}
