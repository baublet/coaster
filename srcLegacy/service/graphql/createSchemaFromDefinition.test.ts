import { printSchema } from "graphql";

import { ServiceType } from "service/types";
import { createSchemaFromDefinition } from "./createSchemaFromDefinition";
import { GraphQLType } from "./types";
import { createResolver } from "./createResolver";
import { createType } from "./createType";
import { notNull } from "./notNull";
import { createUsersAndTodos } from "testHelpers/Todo";

const TestObject = createType({
  type: GraphQLType.OBJECT,
  name: "TestCollection",
  nullable: false,
  description: "This is a test description",
  nodes: {
    totalCount: {
      type: GraphQLType.INT,
      nullable: false
    },
    nodes: {
      type: GraphQLType.INT
    }
  }
});

it("creates a schema", async () => {
  const { User } = await createUsersAndTodos();
  expect(
    printSchema(
      createSchemaFromDefinition({
        name: "name",
        port: 82,
        type: ServiceType.GRAPHQL,
        options: {
          queries: {
            queryWithObjectReturn: createResolver({
              description: "description",
              resolutionType: notNull(TestObject),
              resolver: async () => {
                return null;
              }
            }),
            queryWithPrimitiveReturn: {
              resolutionType: {
                type: GraphQLType.BOOLEAN
              },
              resolver: async () => {}
            },
            queryWithListReturn: {
              resolutionType: {
                type: GraphQLType.ARRAY_OF,
                value: {
                  type: GraphQLType.OBJECT,
                  name: "TestObjectWithinAList",
                  nodes: {
                    testListOfObjects: {
                      type: GraphQLType.ID,
                      nullable: false
                    },
                    testModel: {
                      type: GraphQLType.MODEL,
                      modelFactory: User,
                      description: "Some user here"
                    },
                    enum: {
                      type: GraphQLType.ENUM,
                      name: "SomeTestEnum",
                      values: {
                        ENUM: "description",
                        ANOTHER: "another description"
                      }
                    }
                  }
                }
              },
              resolver: async () => {}
            },
            queryWithCollectionReturn: createResolver({
              description: "Resolver description",
              resolutionType: {
                type: GraphQLType.CONNECTION,
                description: "Connection description",
                name: "TestingConnections",
                of: {
                  type: GraphQLType.STRING
                }
              },
              resolver: async () => {
                return null;
              }
            })
          }
        }
      })
    )
  ).toMatchSnapshot();
});
