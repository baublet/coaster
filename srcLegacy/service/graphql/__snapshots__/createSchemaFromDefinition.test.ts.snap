// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a schema 1`] = `
"type modelFactoryModelConnection {
  totalCount: Int!
  nodes: [Settings_0!]
}

type Query {
  queryWithObjectReturn: TestCollection!
  queryWithPrimitiveReturn: Boolean
  queryWithListReturn: [TestObjectWithinAList]
  queryWithCollectionReturn: TestingConnections
}

type Settings_0 {
  id: String
  bio: String
}

enum SomeTestEnum {
  \\"\\"\\"description\\"\\"\\"
  ENUM

  \\"\\"\\"another description\\"\\"\\"
  ANOTHER
}

\\"\\"\\"This is a test description\\"\\"\\"
type TestCollection {
  totalCount: Int!
  nodes: Int
}

\\"\\"\\"Connection description\\"\\"\\"
type TestingConnections {
  totalCount: Int!
  nodes: [String]!
}

type TestObjectWithinAList {
  testListOfObjects: ID!
  testModel: UserWithTodos_0
  enum: SomeTestEnum
}

\\"\\"\\"Some user here\\"\\"\\"
type UserWithTodos_0 {
  id: String
  name: String
  settings: modelFactoryModelConnection
  todos: modelFactoryModelConnection
  todoGroups: modelFactoryModelConnection
}
"
`;
